tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php

```diff
@@ -882,4 +882,55 @@ public function testJoinWithNonUniqueAliasThrowsException()
 
         $qb->getSQL();
     }
+
+    public function testNormalizedParamNames()
+    {
+        $qb = new QueryBuilder($this->conn);
+
+        $qb->setParameter('foo', 'foo');
+        $this->assertSame('foo', $qb->getParameter('foo'));
+        $this->assertSame('foo', $qb->getParameter(':foo'));
+
+        $qb->setParameter(':bar', 2, \PDO::PARAM_INT);
+        $this->assertSame(2, $qb->getParameter('bar'));
+        $this->assertSame(2, $qb->getParameter(':bar'));
+
+        $this->assertSame(array(
+            'foo' => 'foo',
+            'bar' => 2,
+        ), $qb->getParameters());
+
+        $this->assertSame(array(
+            'bar' => \PDO::PARAM_INT,
+        ), $qb->getParameterTypes());
+
+
+        $qb->setParameters(array(
+            ':foo' => 42,
+            'bar' => 'bar',
+        ), array(
+            ':foo' => \PDO::PARAM_INT,
+        ));
+
+        $this->assertSame(array(
+            'foo' => 42,
+            'bar' => 'bar',
+        ), $qb->getParameters());
+
+        $this->assertSame(array(
+            'foo' => \PDO::PARAM_INT,
+        ), $qb->getParameterTypes());
+    }
+
+    public function testRejectsInvalidParamNames()
+    {
+        $qb = new QueryBuilder($this->conn);
+
+        $this->setExpectedException(
+            'Doctrine\DBAL\Query\QueryException',
+            "The given name '1a' is invalid."
+        );
+
+        $qb->setParameter('1a', 'one a');
```

Comment:
The plan here is to use PDO's validation strategy http://stackoverflow.com/a/5810058/3779 but allow string integers like "1" to pass through.